gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -387,13 +387,7 @@ private GoogleCloudStorageImpl(
 
     this.httpRequestInitializer = this.storage.getRequestFactory().getInitializer();
 
-    this.metricsRecorder =
-        MetricsSink.CLOUD_MONITORING == this.storageOptions.getMetricsSink()
-            ? CloudMonitoringMetricsRecorder.create(
-                options.getProjectId(),
-                new CredentialAdapter(
-                    ((RetryHttpInitializer) httpRequestInitializer).getCredential()))
-            : new NoOpMetricsRecorder();
+    this.metricsRecorder = getMetricsRecorder(options, credential, this.httpRequestInitializer);
```

Comment:
In `getMetricRecorder` we are giving priority to `credential` over the credential stored in `httpRequestInitializer` (extracted from Storage object).
This is a behaviour change and might be breaking change for customer. How we are validating that this is a correct thing to do?

One can see a similar pattern for When gRPC is enabled where passed(passed in constructor call) in credentials are used.I can see the cain of priority as `credentials >credential > httpRequestInitializer`.

Why not extract the logic of which credentials gets the priority and and just use that credential in rest of the function instead of deciding the priority at two different places?

I do understand that in `getMetricsRecorder` too we are maintaining the priority as `credential> httpRequestInitializer` but having this same logic segregated at various place will create issue in maintaining consistency.

I will let you take the call.

Reply:
The priority is the same as the existing code. 

We have refactored this piece of code in the master branch. If we have to refactor this, it will be a bigger change, which is not related to this bug. It can be taken up as a separate change if you feel strongly about it.